source/hp/dof_handler.cc

```diff
@@ -1745,55 +1724,169 @@ namespace hp
+
+              const std::vector<types::subdomain_id> &true_subdomain_ids =
+                shared_tria->get_true_subdomain_ids_of_cells();
+
+              for (unsigned int index = 0; cell != endc; ++cell, ++index)
+                {
+                  saved_subdomain_ids[index] = cell->subdomain_id();
+                  cell->set_subdomain_id(true_subdomain_ids[index]);
+                }
+
+              // Make sure every processor knows the active_fe_indices
+              // on both its own cells and all ghost cells.
+              dealii::internal::hp::DoFHandlerImplementation::Implementation::
+                communicate_active_fe_indices(*this);
+            }
+
+        // Store active_fe_index information for all cells that will be
+        // affected by refinement/coarsening.
+        for (const auto &cell : active_cell_iterators())
           {
-            // Look if the cell got children during refinement by
-            // checking whether it has children now but didn't have
-            // children before refinement (the has_children array is
-            // set in pre-refinement action)
-            //
-            // Note: Although one level is added to the DoFHandler
-            // levels, when the triangulation got one, for the buffer
-            // has_children this new level is not required, because
-            // the cells on the finest level never have
-            // children. Hence cell->has_children () will always
-            // return false on that level, which would cause shortcut
-            // evaluation of the following expression. Thus an index
-            // error in has_children should never occur.
-            if (cell->has_children() &&
-                !(*has_children[cell->level()])[cell->index()])
+            if (cell->refine_flag_set())
               {
-                // Set active_fe_index in children to the same value
-                // as in the parent cell. we can't access the
-                // active_fe_index in the parent cell any more through
-                // cell->active_fe_index() since that function is not
-                // allowed for inactive cells, but we can access this
-                // information from the DoFLevels directly
-                //
-                // we don't have to set the active_fe_index for ghost
-                // cells -- these will be exchanged automatically
-                // upon distribute_dofs()
-                for (unsigned int i = 0; i < cell->n_children(); ++i)
-                  if (cell->child(i)->is_locally_owned())
-                    cell->child(i)->set_active_fe_index(
-                      levels[cell->level()]->active_fe_index(cell->index()));
+                // Store the active_fe_index of each cell that will be refined
+                // to and distribute it later on its children.
+                refined_cells_fe_index.push_back(
+                  {cell->id(), cell->active_fe_index()});
```

Comment:
this refers to the comment I had on the declaration of the variable: I think a more efficient way would be to just store the `cell` iterator, rather than its id. That's because you only work with iterators that are guaranteed to continue to exist after the refinement operation.